src/compositor.rs

```diff
@@ -275,51 +324,54 @@ where
                     if let Some(data) = surface.data::<U>() {
                         let data = data.surface_data();
                         let inner = data.inner.lock().unwrap();
-                        let current = data.scale_factor.load(Ordering::Relaxed);
-                        let factor = match inner
-                            .outputs
-                            .iter()
-                            .filter_map(|output| {
-                                output.data::<OutputData>().map(OutputData::scale_factor)
-                            })
-                            .reduce(i32::max)
-                        {
-                            None => return,
-                            Some(factor) if factor == current => return,
-                            Some(factor) => factor,
-                        };
-
-                        data.scale_factor.store(factor, Ordering::Relaxed);
-                        drop(inner);
-                        state.scale_factor_changed(conn, qh, &surface, factor);
+                        dispatch_surface_state_updates(state, conn, qh, &surface, data, inner);
                     }
                 }
             })
         });
 
-        // Compute the new max of the scale factors for all outputs this surface is displayed on.
-        let current = data.scale_factor.load(Ordering::Relaxed);
-
-        let factor = match inner
-            .outputs
-            .iter()
-            .filter_map(|output| output.data::<OutputData>().map(OutputData::scale_factor))
-            .reduce(i32::max)
-        {
-            // If no scale factor is found, because the surface has left its only output, do not
-            // change the scale factor.
-            None => return,
-            Some(factor) if factor == current => return,
-            Some(factor) => factor,
-        };
+        dispatch_surface_state_updates(state, conn, qh, surface, data, inner);
+    }
+}
 
-        data.scale_factor.store(factor, Ordering::Relaxed);
+fn dispatch_surface_state_updates<D, U>(
```

Comment:
Because this method is only called if the version is less than 6, will the surface transform ever be anything other than `wl_output::Transform::Normal`? It seems to default to `wl_output::Transform::Normal` and this method will only set it to `wl_output::Transform::Normal` as far as I can tell from the reduce step, so I'm not sure it needs to do anything with transform. Or have I missed something?

Reply:
>Because this method is only called if the version is less than 6, will the surface transform ever be anything other than wl_output::Transform::Normal? It seems to default to wl_output::Transform::Normal and this method will only set it to wl_output::Transform::Normal as far as I can tell from the reduce step, so I'm not sure it needs to do anything with transform. Or have I missed something?

It might be really confusing, but reduce doesn't apply when you have 1 element. So for more than 2 elements in array, it'll map to `Normal`, but for exactly one output it'll be the corrent transform.